src/sanitizer.js

```diff
@@ -241,3 +249,83 @@ export function isValidAttr(attrName, attrValue) {
 
   return true;
 }
+
+/**
+ * Resolves the attribute value. The main purpose is to rewrite URLs as
+ * described in `resolveUrlAttr`.
+ * @param {string} tagName
+ * @param {string} attrName
+ * @param {string} attrValue
+ * @return {string}
+ */
+function resolveAttrValue(tagName, attrName, attrValue) {
+  if (attrName == 'src' || attrName == 'href' || attrName == 'srcset') {
+    return resolveUrlAttr(tagName, attrName, attrValue, window.location);
+  }
+  return attrValue;
+}
+
+/**
+ * Rewrites the URL attribute values. URLs are rewritten as following:
+ * - If URL is absolute, it is not rewritten
+ * - If URL is relative, it's rewritten as absolute against the source origin
+ * - If resulting URL is a `http:` URL and it's for image, the URL is rewritten
+ *   again to be served with AMP Cache (cdn.ampproject.org).
+ *
+ * @param {string} tagName
+ * @param {string} attrName
+ * @param {string} attrValue
+ * @param {!Location} windowLocation
+ * @return {string}
+ * @private Visible for testing.
+ */
+export function resolveUrlAttr(tagName, attrName, attrValue, windowLocation) {
+  const isProxy = isProxyOrigin(windowLocation);
```

Comment:
All these string optimizations are going to be a perf problem. We can optimize later, but it won't be pretty.